### PR TITLE
Add info to let named types quickly return 'null' when they have no primary constructor

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Traces show several places in the compiler where code wants to determine if a type has a primary constructor to choose a path to take.  Currently, this can only be computed by actually generating all members, then looking up if there's a primary constructor in them.  This PR allows hte question to be answered (in the common case, where there is no primary constructor) by storing a bit in the flags of the type.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 1]
```diff
@@ -30,15 +30,16 @@ private struct Flags
         {
             // We current pack everything into one 32-bit int; layout is given below.
             //
-            // |             |ss|vvv|zzzz|f|d|yy|wwwwww|
+            // |           p|ss|vvv|zzzz|f|d|yy|wwwwww|
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:33]:
```suggestion
            // |            p|ss|vvv|zzzz|f|d|yy|wwwwww|
```

File: src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs [hunk 2]
```diff
@@ -614,10 +614,11 @@ private SingleNamespaceOrTypeDeclaration VisitTypeDeclaration(TypeDeclarationSyn
             }
 
             // A type with parameters at least has a primary constructor
-            bool hasPrimaryCtor = node is TypeDeclarationSyntax { ParameterList: { } };
+            var hasPrimaryCtor = node.ParameterList != null;
             if (hasPrimaryCtor)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs:618]: >if (hasPrimaryCtor) [](http://example.com/codeflow?start=12&length=19)

Probably shouldn't enter this ```if``` for interfaces.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs:618]: I will clean this up separately.

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs:618]: sgtm.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 5)

**Comment by CyrusNajmabadi** [comment]
@RikkiGibson ptal.

**Review by cston — APPROVED** [review]

**Pull request merged**